File: L1S/v3-ocaml/code/ocaml/src/regalloc.ml
open Ir

module VRegMap = Map.Make(Int)

type phys_reg =
  | Rax | Rcx | Rdx | Rbx | Rsi | Rdi
  | R8 | R9 | R10 | R11 | R12 | R13 | R14 | R15

let all_regs = [
  Rax; Rcx; Rdx; Rbx; Rsi; Rdi;
  R8; R9; R10; R11; R12; R13; R14; R15;
]

let arg_regs = [Rdi; Rsi; Rdx; Rcx; R8; R9]

type live_interval = {
  vreg: int;
  start_: int;
  end_: int;
}

type allocation = {
  reg_map: phys_reg VRegMap.t;
  spill_slots: int VRegMap.t;
  stack_size: int;
}

type allocator = unit

let create () = ()

let compute_live_intervals func =
  let defs = ref VRegMap.empty in
  let uses = ref VRegMap.empty in
  let instr_counter = ref 0 in

  List.iter (fun block ->
    List.iter (fun instr ->
      let u = uses_of_instr instr in
      List.iter (fun v ->
        let existing = match VRegMap.find_opt v !uses with
          | Some l -> l | None -> []
        in
        uses := VRegMap.add v (existing @ [!instr_counter]) !uses
      ) u;
      (match dest_of_instr instr with
       | Some d -> defs := VRegMap.add d !instr_counter !defs
       | None -> ());
      instr_counter := !instr_counter + 1
    ) block.instrs;
    (match block.terminator with
     | Ret (Some v) ->
       let existing = match VRegMap.find_opt v !uses with
         | Some l -> l | None -> []
       in
       uses := VRegMap.add v (existing @ [!instr_counter]) !uses
     | BrCond (v, _, _) ->
       let existing = match VRegMap.find_opt v !uses with
         | Some l -> l | None -> []
       in
       uses := VRegMap.add v (existing @ [!instr_counter]) !uses
     | _ -> ());
    instr_counter := !instr_counter + 1
  ) func.f_blocks;

  let intervals = ref [] in
  VRegMap.iter (fun vreg def ->
    let last_use = match VRegMap.find_opt vreg !uses with
      | Some l -> List.fold_left max def l
      | None -> def
    in
    intervals := { vreg; start_ = def; end_ = last_use } :: !intervals
  ) !defs;
  VRegMap.iter (fun vreg use_list ->
    if not (VRegMap.mem vreg !defs) then begin
      let last_use = List.fold_left max 0 use_list in
      intervals := { vreg; start_ = 0; end_ = last_use } :: !intervals
    end
  ) !uses;

  List.sort (fun a b -> compare a.start_ b.start_) !intervals

let get_spill_slot counter =
  let s = !counter in
  counter := !counter + 1;
  s

let expire_intervals current_start active reg_map =
  let (keep, _) = List.partition (fun (e, v, r) ->
    if e <= current_start then begin
      reg_map := VRegMap.remove v !reg_map;
      false
    end else true
  ) !active in
  active := keep

let allocate func =
  let intervals = compute_live_intervals func in
  let sorted = List.sort (fun a b -> compare a.start_ b.start_) intervals in
  let reg_map = ref VRegMap.empty in
  let spill_slots = ref VRegMap.empty in
  let next_spill_slot = ref 0 in
  let active = ref [] in

  List.iter (fun interval ->
    expire_intervals interval.start_ active reg_map;

    if List.length !active < List.length all_regs then begin
      let reg = List.nth all_regs (List.length !active) in
      reg_map := VRegMap.add interval.vreg reg !reg_map;
      active := (interval.end_, interval.vreg, reg) :: !active
    end else begin
      let spill_candidate =
        let (max_end, max_vreg, _) = List.fold_left (fun (me, mv, _) (e, v, _) ->
          if e > me then (e, v, true) else (me, mv, true)
        ) (0, 0, false) !active
        in max_vreg
      in
      let candidate_interval = List.find (fun i -> i.vreg = spill_candidate) intervals in
      if candidate_interval.end_ > interval.end_ then begin
        let reg = VRegMap.find spill_candidate !reg_map in
        reg_map := VRegMap.remove spill_candidate !reg_map;
        let slot = get_spill_slot next_spill_slot in
        spill_slots := VRegMap.add spill_candidate slot !spill_slots;
        reg_map := VRegMap.add interval.vreg reg !reg_map;
        active := List.filter (fun (_, v, _) -> v <> spill_candidate) !active;
        active := (interval.end_, interval.vreg, reg) :: !active
      end else begin
        let slot = get_spill_slot next_spill_slot in
        spill_slots := VRegMap.add interval.vreg slot !spill_slots
      end
    end
  ) sorted;

  (* Any remaining unassigned vregs get a spill slot *)
  List.iter (fun interval ->
    if not (VRegMap.mem interval.vreg !reg_map) &&
       not (VRegMap.mem interval.vreg !spill_slots) then begin
      let slot = get_spill_slot next_spill_slot in
      spill_slots := VRegMap.add interval.vreg slot !spill_slots
    end
  ) intervals;

  { reg_map = !reg_map; spill_slots = !spill_slots; stack_size = !next_spill_slot * 8 }

let phys_name = function
  | Rax -> "rax" | Rcx -> "rcx" | Rdx -> "rdx" | Rbx -> "rbx"
  | Rsi -> "rsi" | Rdi -> "rdi"
  | R8 -> "r8" | R9 -> "r9" | R10 -> "r10" | R11 -> "r11"
  | R12 -> "r12" | R13 -> "r13" | R14 -> "r14" | R15 -> "r15"

let arg_reg_name i =
  if i < List.length arg_regs then phys_name (List.nth arg_regs i)
  else "rdi"

File: L1S/v3-rust/code/rust/src/regalloc.rs
use crate::ir::*;
use std::collections::HashMap;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum PhysReg {
    Rax,
    Rcx,
    Rdx,
    Rbx,
    Rsi,
    Rdi,
    R8,
    R9,
    R10,
    R11,
    R12,
    R13,
    R14,
    R15,
}

pub const ALL_REGS: &[PhysReg] = &[
    PhysReg::Rax,
    PhysReg::Rcx,
    PhysReg::Rdx,
    PhysReg::Rbx,
    PhysReg::Rsi,
    PhysReg::Rdi,
    PhysReg::R8,
    PhysReg::R9,
    PhysReg::R10,
    PhysReg::R11,
    PhysReg::R12,
    PhysReg::R13,
    PhysReg::R14,
    PhysReg::R15,
];

pub const CALLEE_SAVED: &[PhysReg] = &[
    PhysReg::Rbx,
    PhysReg::R12,
    PhysReg::R13,
    PhysReg::R14,
    PhysReg::R15,
];

pub const ARG_REGS: &[PhysReg] = &[
    PhysReg::Rdi,
    PhysReg::Rsi,
    PhysReg::Rdx,
    PhysReg::Rcx,
    PhysReg::R8,
    PhysReg::R9,
];

#[derive(Clone, Debug)]
pub struct LiveInterval {
    pub vreg: VReg,
    pub start: usize,
    pub end: usize,
}

#[derive(Clone, Debug)]
pub struct Allocation {
    pub reg_map: HashMap<VReg, PhysReg>,
    pub spill_slots: HashMap<VReg, usize>,
    pub stack_size: usize,
}

pub struct RegAllocator;

impl RegAllocator {
    pub fn new() -> Self {
        RegAllocator
    }

    pub fn allocate(&self, func: &IrFunction) -> Allocation {
        let intervals = self.compute_live_intervals(func);
        let mut sorted: Vec<LiveInterval> = intervals.clone();
        sorted.sort_by_key(|i| i.start);

        let mut reg_map: HashMap<VReg, PhysReg> = HashMap::new();
        let mut spill_slots: HashMap<VReg, usize> = HashMap::new();
        let mut next_spill_slot: usize = 0;

        let mut active: Vec<(usize, VReg, PhysReg)> = Vec::new();

        for interval in &sorted {
            self.expire_intervals(interval.start, &mut active, &mut reg_map);

            if active.len() < ALL_REGS.len() {
                let reg = ALL_REGS[active.len()];
                reg_map.insert(interval.vreg, reg);
                active.push((interval.end, interval.vreg, reg));
            } else {
                let spill_candidate = active
                    .iter()
                    .max_by_key(|(end, _, _)| *end)
                    .unwrap()
                    .1;
                let candidate_interval = intervals.iter().find(|i| i.vreg == spill_candidate).unwrap();

                if candidate_interval.end > interval.end {
                    // Spill candidate (farther end) instead
                    let reg = reg_map[&spill_candidate];
                    reg_map.remove(&spill_candidate);
                    let slot = self.get_spill_slot(&mut next_spill_slot);
                    spill_slots.insert(spill_candidate, slot);
                    reg_map.insert(interval.vreg, reg);
                    active.retain(|(_, v, _)| *v != spill_candidate);
                    active.push((interval.end, interval.vreg, reg));
                } else {
                    // Spill current interval
                    let slot = self.get_spill_slot(&mut next_spill_slot);
                    spill_slots.insert(interval.vreg, slot);
                }
            }
        }

        // Any vregs not assigned to a register get a spill slot
        for interval in &intervals {
            if !reg_map.contains_key(&interval.vreg) && !spill_slots.contains_key(&interval.vreg) {
                let slot = self.get_spill_slot(&mut next_spill_slot);
                spill_slots.insert(interval.vreg, slot);
            }
        }

        Allocation {
            reg_map,
            spill_slots,
            stack_size: next_spill_slot * 8,
        }
    }

    fn get_spill_slot(&self, counter: &mut usize) -> usize {
        let slot = *counter;
        *counter += 1;
        slot
    }

    fn expire_intervals(
        &self,
        current_start: usize,
        active: &mut Vec<(usize, VReg, PhysReg)>,
        reg_map: &mut HashMap<VReg, PhysReg>,
    ) {
        active.retain(|(end, vreg, reg)| {
            if *end <= current_start {
                reg_map.remove(vreg);
                false
            } else {
                true
            }
        });
    }

    fn compute_live_intervals(&self, func: &IrFunction) -> Vec<LiveInterval> {
        let mut defs: HashMap<VReg, usize> = HashMap::new();
        let mut uses: HashMap<VReg, Vec<usize>> = HashMap::new();
        let mut instr_counter: usize = 0;

        for block in &func.blocks {
            for instr in &block.instrs {
                for u in instr.uses() {
                    uses.entry(u).or_default().push(instr_counter);
                }
                if let Some(d) = instr.dest() {
                    defs.entry(d).or_insert(instr_counter);
                }
                instr_counter += 1;
            }
            match &block.terminator {
                IrTerminator::Ret(Some(v)) => {
                    uses.entry(*v).or_default().push(instr_counter);
                }
                IrTerminator::BrCond(v, _, _) => {
                    uses.entry(*v).or_default().push(instr_counter);
                }
                _ => {}
            }
            instr_counter += 1;
        }

        let mut intervals = Vec::new();
        for (&vreg, &def) in &defs {
            let last_use = uses.get(&vreg).and_then(|u| u.iter().max()).copied().unwrap_or(def);
            intervals.push(LiveInterval {
                vreg,
                start: def,
                end: last_use,
            });
        }

        for (&vreg, uses_list) in &uses {
            if !defs.contains_key(&vreg) {
                let last_use = uses_list.iter().max().copied().unwrap_or(0);
                intervals.push(LiveInterval {
                    vreg,
                    start: 0,
                    end: last_use,
                });
            }
        }

        intervals
    }
}
